nifi-nar-bundles/nifi-kafka-bundle/nifi-kafka-1-0-processors/src/main/java/org/apache/nifi/record/sink/kafka/KafkaRecordSink_1_0.java

```diff
@@ -232,64 +238,77 @@ public void onEnabled(final ConfigurationContext context) throws InitializationE
     public WriteResult sendData(final RecordSet recordSet, final Map<String, String> attributes, final boolean sendZeroResults) throws IOException {
 
         try {
-            WriteResult writeResult;
             final RecordSchema writeSchema = getWriterFactory().getSchema(null, recordSet.getSchema());
             final ByteArrayOutputStream baos = new ByteArrayOutputStream();
             final ByteCountingOutputStream out = new ByteCountingOutputStream(baos);
             int recordCount = 0;
             try (final RecordSetWriter writer = getWriterFactory().createWriter(getLogger(), writeSchema, out, attributes)) {
-                writer.beginRecordSet();
                 Record record;
                 while ((record = recordSet.next()) != null) {
```

Comment:
It's been a while since I looked at this but the intent was to send individual messages rather than all records in one message. Can this still cause heap exhaustion if we don't ack them until they're all sent? What would the property be, maybe Messages per Batch?

Reply:
@mattyb149 sorry I think I read the code wrong :) Here we are using `baos.reset()` to clear the buffer in between in each record. Ignore this comment.